### ISSUE TITLE
Segfault with http2 with bad connection

### ISSUE THREAD

**Issue by ThomasObenaus** [curl/curl#2688]
### I did this
Application that communicates using http2 (nghttp2) with our backend.
In case the connection is bad (mobile phone with edge/gsm) the application crashes each time with a segfault and the same back-trace.

### Precondition
* Bad connection (low bandwidth) with over mobile phone whose datavolume is empty (fallback to 2G).
* No connection to the internet (firewall closed).
* Application starts requesting (to reproduce the problem) a lot of requests. 
* dnsmask already complains about too many requests in parallel (max of 150 reached).
* Opening the fire wall.
* Waiting for ~1 min --> segfault.

### Curl errors
* When FW is closed: "Cloud not connect"
* When FW is open: "Timeout reached"

### I expected the following
no segfault

### curl/libcurl version
Tested with 7.54.0 and 7.60

### Libs
nghttp2 1.17.0
c-ares 1.13.9

**Comment by ThomasObenaus** [comment]
## Backtrace
`(gdb) bt
#0 0x0000007f935e0988 in __GI__IO_fwrite (buf=..., size=..., count=..., fp=...) at iofwrite.c:37
#1 0x0000007f956708a8 in showit (data=..., type=..., ptr=..., size=...) at sendf.c:826
#2 0x0000007f95670c18 in Curl_debug (data=..., type=..., ptr=..., size=..., conn=...) at sendf.c:878
#3 0x0000007f95670d50 in Curl_infof (data=..., fmt=...) at sendf.c:234
#4 0x0000007f956a1f9c in on_frame_not_send (session=..., frame=..., lib_error_code=..., userp=...) at http2.c:722
#5 0x0000007f95636380 in nghttp2_session_mem_send_internal (session=..., data_ptr=..., fast_cb=...)
at /home/thomaso/work/projects/bob_nav/dev/src/libs/ext-nghttp2/1/workspace/nghttp2-1.17.0/lib/nghttp2_session.c:2914
#6 0x0000007f95636c04 in nghttp2_session_send (session=...)
at /home/thomaso/work/projects/bob_nav/dev/src/libs/ext-nghttp2/1/workspace/nghttp2-1.17.0/lib/nghttp2_session.c:3215
#7 0x0000007f956a29ec in h2_session_send (data=..., h2=...) at http2.c:1379
#8 0x0000007f956a405c in http2_recv (conn=..., sockindex=..., mem=..., len=..., err=...) at http2.c:1559
#9 0x0000007f95670a64 in Curl_read (conn=..., sockfd=..., buf=..., sizerequested=..., n=...) at sendf.c:747
#10 0x0000007f95684764 in readwrite_data (comeback=..., done=..., didwhat=..., k=..., conn=..., data=...)
at transfer.c:434
#11 Curl_readwrite (conn=..., data=..., done=..., comeback=...) at transfer.c:1102
#12 0x0000007f9568ebac in multi_runsingle (multi=..., now=..., data=...) at multi.c:1908
#13 0x0000007f9568f5a0 in curl_multi_perform (multi=..., running_handles=...) at multi.c:2180
#14 0x00000000017ffeec in tsd::nav::cloud::comm::client::HTTPRequestHandler::main() ()
#15 0x00000000027f5ffc in tsd::nav::cloud::comm::client::HTTPRequestHandler::run() ()
#16 0x0000007f939fc510 in tsd::common::system::Thread::startImpl (arg=...)
at /home/thomaso/work/projects/bob_nav/dev/src/system/tsd-common/1/workspace/tsd.common/src/tsd/common/system/Thread.cpp:310
#17 0x0000007f93a00040 in tsd::common::system::os::thread::pseudoStartThread (arg=...)
at /home/thomaso/work/projects/bob_nav/dev/src/system/tsd-common/1/workspace/tsd.common/src/tsd/common/system/posix/POSIXThread.cpp:367
#18 0x0000007f93955030 in start_thread (arg=...) at pthread_create.c:331
#19 0x0000007f936464c0 in thread_start () at ../sysdeps/unix/sysv/linux/aarch64/clone.S:89`

**Comment by bagder** [comment]
That's a 7.54.0 backtrace, right? The line numbers seems way off from 7.60.0... Did you build with `DEBUG_HTTP2` defined?

So what is the problem in the showit() function? The blanked out arguments make it hard to guess.

**Comment by ThomasObenaus** [comment]
# Current state of investigation
* The guess is that after we got back the easy_handle from `curl_multi_info_read` processed it and then freed the easy_handle with `easy_cleanup` there still seams to come traffic for this easy_handle.

**Comment by ThomasObenaus** [comment]
Yes that's the bt from 7.54.0.
In showit() it seams that the easy_handle (data) is damaged.

**Comment by ThomasObenaus** [comment]
* Even at level of Curl_readwrite (f 11) the connection having a pointer to a stream-map referencing another easy_handle (0x7f08776438) than the easy_handle (0x7f08c62258) that owns this connection.
* The first easy_handle is invalid (magic should be 0xc0dedbad).
(gdb) p data->magic 
$6 = 0x877b330
* It seams that after curl_easy_cleanup was called on the easy_handle (0x7f08776438) it is still part of the nghttp2-stream. Thus when the stream is processed in the next loop it crashes.
* The question here is: Who adds the "wrong" easy_handle (0x7f08776438) as stream_user_data to the "good" easy_handle (0x7f08c62258)

**Comment by bagder** [comment]
Since I can't reproduce I can only offer general guidance as how to go ahead and debug this:

- build with git master, we've already fixed a few problems with removed easy handles from the multi handle
- define `DEBUG_HTTP2` and check if you actually get frames from a stream that is already done/removed when you see this problem - which I think it might be
- if you get frames from a stream that was owned by an easy handle that is now removed, can you figure out why `Curl_http2_done` was not called on it (and  subsequently `nghttp2_session_set_stream_user_data` cleared the association from frame id to easy handle), or whatever it is that makes libcurl extract a "dead" easy handle for an incoming h2 frame

**Comment by ThomasObenaus** [comment]
Thanks for the quick feedback.
I'll rebuild it with your hints and come back as soon as possible with a status update.

**Comment by ThomasObenaus** [comment]
### Update
I tried it with 7.60.0 and get now the same bt like mentioned in https://github.com/curl/curl/issues/2674 for one time and another one at the second try.

#### BT 1
```
#0  h2_session_send (data=..., h2=...) at http2.c:1470
#1  0x0000007fbb9564e8 in h2_process_pending_input (conn=..., httpc=..., err=...) at http2.c:1303
#2  0x0000007fbb95667c in http2_connisdead (conn=...) at http2.c:219
#3  http2_conncheck (check=..., checks_to_perform=...) at http2.c:236
#4  0x0000007fbb92c760 in extract_if_dead (conn=..., data=...) at url.c:973
#5  0x0000007fbb92c7ec in call_extract_if_dead (conn=..., param=...) at url.c:1003
#6  0x0000007fbb954914 in Curl_conncache_foreach (data=..., connc=..., param=..., func=...) at conncache.c:382
#7  0x0000007fbb9303d8 in prune_dead_connections (data=...) at url.c:1025
#8  create_conn (async=..., in_connect=..., data=...) at url.c:4382
#9  Curl_connect (data=..., in_connect=..., asyncp=..., protocol_done=...) at url.c:4660
#10 0x0000007fbb940db0 in multi_runsingle (multi=..., now=..., data=...) at multi.c:1421
#11 0x0000007fbb941b0c in curl_multi_perform (multi=..., running_handles=...) at multi.c:2159
#12 0x00000000017ffe10 in tsd::nav::cloud::comm::client::HTTPRequestHandler::main() ()
#13 0x00000000027f500c in tsd::nav::cloud::comm::client::HTTPRequestHandler::run() ()
#14 0x0000007fb9cb0510 in tsd::common::system::Thread::startImpl (arg=...)
    at /home/thomaso/work/projects/bob_c_ares/dev/src/system/tsd-common/1/workspace/tsd.common/src/tsd/common/system/Thread.cpp:310
#15 0x0000007fb9cb4040 in tsd::common::system::os::thread::pseudoStartThread (arg=...)
    at /home/thomaso/work/projects/bob_c_ares/dev/src/system/tsd-common/1/workspace/tsd.common/src/tsd/common/system/posix/POSIXThread.cpp:367
#16 0x0000007fb9c09030 in start_thread (arg=...) at pthread_create.c:331
#17 0x0000007fb98fa4c0 in thread_start () at ../sysdeps/unix/sysv/linux/aarch64/clone.S:89
```

#### BT 2 
```
#0  0x0000007f88db0504 in Curl_multi_connchanged (multi=...) at multi.c:1111
#1  0x0000007f88dc79cc in on_frame_recv (session=..., frame=..., userp=...) at http2.c:586
#2  0x0000007f88d5b620 in session_call_on_frame_received (frame=..., session=...)
    at /home/thomaso/work/projects/bob_c_ares/dev/src/libs/ext-nghttp2/1/workspace/nghttp2-1.17.0/lib/nghttp2_session.c:3271
#3  nghttp2_session_on_settings_received (session=..., frame=..., noack=...)
    at /home/thomaso/work/projects/bob_c_ares/dev/src/libs/ext-nghttp2/1/workspace/nghttp2-1.17.0/lib/nghttp2_session.c:4367
#4  0x0000007f88d5e7a8 in session_process_settings_frame (session=...)
    at /home/thomaso/work/projects/bob_c_ares/dev/src/libs/ext-nghttp2/1/workspace/nghttp2-1.17.0/lib/nghttp2_session.c:4507
#5  nghttp2_session_mem_recv (session=..., in=..., inlen=...)
    at /home/thomaso/work/projects/bob_c_ares/dev/src/libs/ext-nghttp2/1/workspace/nghttp2-1.17.0/lib/nghttp2_session.c:6161
#6  0x0000007f88dc74c0 in h2_process_pending_input (conn=..., httpc=..., err=...) at http2.c:1279
#7  0x0000007f88dc767c in http2_connisdead (conn=...) at http2.c:219
#8  http2_conncheck (check=..., checks_to_perform=...) at http2.c:236
#9  0x0000007f88d9d760 in extract_if_dead (conn=..., data=...) at url.c:973
#10 0x0000007f88d9d7ec in call_extract_if_dead (conn=..., param=...) at url.c:1003
#11 0x0000007f88dc5914 in Curl_conncache_foreach (data=..., connc=..., param=..., func=...) at conncache.c:382
#12 0x0000007f88da13d8 in prune_dead_connections (data=...) at url.c:1025
#13 create_conn (async=..., in_connect=..., data=...) at url.c:4382
#14 Curl_connect (data=..., in_connect=..., asyncp=..., protocol_done=...) at url.c:4660
#15 0x0000007f88db1db0 in multi_runsingle (multi=..., now=..., data=...) at multi.c:1421
#16 0x0000007f88db2b0c in curl_multi_perform (multi=..., running_handles=...) at multi.c:2159
#17 0x00000000017ffe10 in tsd::nav::cloud::comm::client::HTTPRequestHandler::main() ()
#18 0x00000000027f500c in tsd::nav::cloud::comm::client::HTTPRequestHandler::run() ()
#19 0x0000007f87121510 in tsd::common::system::Thread::startImpl (arg=...)
    at /home/thomaso/work/projects/bob_c_ares/dev/src/system/tsd-common/1/workspace/tsd.common/src/tsd/common/system/Thread.cpp:310
#20 0x0000007f87125040 in tsd::common::system::os::thread::pseudoStartThread (arg=...)
    at /home/thomaso/work/projects/bob_c_ares/dev/src/system/tsd-common/1/workspace/tsd.common/src/tsd/common/system/posix/POSIXThread.cpp:367
#21 0x0000007f8707a030 in start_thread (arg=...) at pthread_create.c:331
#22 0x0000007f86d6b4c0 in thread_start () at ../sysdeps/unix/sysv/linux/aarch64/clone.S:89
```

**Comment by ThomasObenaus** [comment]
Next step: use master from libcurl

**Comment by ThomasObenaus** [comment]
Still a segfault, even with master from lib-curl:
```
#0  on_stream_close (session=..., stream_id=..., error_code=..., userp=...) at http2.c:842
#1  0x0000007f8a279c38 in nghttp2_session_close_stream (session=..., stream_id=..., error_code=...)
    at /home/thomaso/work/projects/bob_c_ares/dev/src/libs/ext-nghttp2/1/workspace/nghttp2-1.17.0/lib/nghttp2_session.c:1162
#2  0x0000007f8a27d3ac in nghttp2_session_mem_send_internal (session=..., data_ptr=..., fast_cb=...)
    at /home/thomaso/work/projects/bob_c_ares/dev/src/libs/ext-nghttp2/1/workspace/nghttp2-1.17.0/lib/nghttp2_session.c:2945
#3  0x0000007f8a27dc04 in nghttp2_session_send (session=...)
    at /home/thomaso/work/projects/bob_c_ares/dev/src/libs/ext-nghttp2/1/workspace/nghttp2-1.17.0/lib/nghttp2_session.c:3215
#4  0x0000007f8a2e81ec in h2_session_send (data=..., h2=...) at http2.c:1476
#5  0x0000007f8a2e9924 in http2_recv (conn=..., sockindex=..., mem=..., len=..., err=...) at http2.c:1656
#6  0x0000007f8a2b82e0 in Curl_read (conn=..., sockfd=..., buf=..., sizerequested=..., n=...) at sendf.c:749
#7  0x0000007f8a2c8db0 in readwrite_data (comeback=..., done=..., didwhat=..., k=..., conn=..., data=...)
    at transfer.c:482
#8  Curl_readwrite (conn=..., data=..., done=..., comeback=...) at transfer.c:1125
#9  0x0000007f8a2d3144 in multi_runsingle (multi=..., now=..., data=...) at multi.c:1914
#10 0x0000007f8a2d3b8c in curl_multi_perform (multi=..., running_handles=...) at multi.c:2181
#11 0x00000000017e6ff0 in tsd::nav::cloud::comm::client::HTTPRequestHandler::main() ()
#12 0x00000000027a3ecc in tsd::nav::cloud::comm::client::HTTPRequestHandler::run() ()
#13 0x0000007f88642510 in tsd::common::system::Thread::startImpl (arg=...)
    at /home/thomaso/work/projects/bob_c_ares/dev/src/system/tsd-common/1/workspace/tsd.common/src/tsd/common/system/Thread.cpp:310
#14 0x0000007f88646040 in tsd::common::system::os::thread::pseudoStartThread (arg=...)
    at /home/thomaso/work/projects/bob_c_ares/dev/src/system/tsd-common/1/workspace/tsd.common/src/tsd/common/system/posix/POSIXThread.cpp:367
#15 0x0000007f8859c030 in start_thread (arg=...) at pthread_create.c:331
#16 0x0000007f8828d4c0 in thread_start () at ../sysdeps/unix/sysv/linux/aarch64/clone.S:89
```

**Comment by ThomasObenaus** [comment]
@bagder: It looks that I can reproduce it with my setup quite good. 
Do you have any more ideas, maybe code locations which I should instrument to track down the issue?

**Comment by bagder** [comment]
That's an incoming stream close. It looks like `stream` is perhaps pointing to already freed memory there? If that's so, can you figure out why/how it was freed already? It looks like the free in `http2_disconnect()` is a  likely suspect but still seems weird... Can you run your code with valgrind? If so, does it help to point out any problem perhaps earlier?

If you have http2 debugging enabled you can see what stream id that ends there. Can you see anything wrong with that? Is there other streams going on at the same time?

> p *session

That's the libnghttp2 handle, I don't think that helps us much because I think the problem is on the curl side.

**Comment by ThomasObenaus** [comment]
### With HTTP2_DEBUG enabled
Now I get an assert at `DEBUGASSERT(httpc->drain_total >= data->state.drain);`
```
#0  0x0000007faa6851e8 in __GI_raise (sig=...) at ../sysdeps/unix/sysv/linux/raise.c:54
#1  0x0000007faa6865dc in __GI_abort () at abort.c:89
#2  0x0000007faa67e94c in __assert_fail_base (fmt=..., assertion=..., file=..., line=..., function=...) at assert.c:92
#3  0x0000007faa67e9e4 in __GI___assert_fail (assertion=..., file=..., line=..., function=...) at assert.c:101
#4  0x0000007fac775194 in http2_handle_stream_close (conn=..., data=..., stream=..., err=...) at http2.c:1370
#5  0x0000007fac775c68 in http2_recv (conn=..., sockindex=..., mem=..., len=..., err=...) at http2.c:1693
#6  0x0000007fac7442e0 in Curl_read (conn=..., sockfd=..., buf=..., sizerequested=..., n=...) at sendf.c:749
#7  0x0000007fac754db0 in readwrite_data (comeback=..., done=..., didwhat=..., k=..., conn=..., data=...)
    at transfer.c:482
#8  Curl_readwrite (conn=..., data=..., done=..., comeback=...) at transfer.c:1125
#9  0x0000007fac75f144 in multi_runsingle (multi=..., now=..., data=...) at multi.c:1914
#10 0x0000007fac75fb8c in curl_multi_perform (multi=..., running_handles=...) at multi.c:2181
#11 0x00000000017e6ff0 in tsd::nav::cloud::comm::client::HTTPRequestHandler::main() ()
#12 0x00000000027a3ecc in tsd::nav::cloud::comm::client::HTTPRequestHandler::run() ()
#13 0x0000007faaace510 in tsd::common::system::Thread::startImpl (arg=...)
    at /home/thomaso/work/projects/bob_c_ares/dev/src/system/tsd-common/1/workspace/tsd.common/src/tsd/common/system/Thread.cpp:310
#14 0x0000007faaad2040 in tsd::common::system::os::thread::pseudoStartThread (arg=...)
    at /home/thomaso/work/projects/bob_c_ares/dev/src/system/tsd-common/1/workspace/tsd.common/src/tsd/common/system/posix/POSIXThread.cpp:367
#15 0x0000007faaa28030 in start_thread (arg=...) at pthread_create.c:331
#16 0x0000007faa7194c0 in thread_start () at ../sysdeps/unix/sysv/linux/aarch64/clone.S:89
```
#### Last Log-Lines
```
16:50:35.040    [Debug]  [0x7f352d6a88] cURL [TEXT] Using Stream ID: e5 (easy handle 0x7f352d6a88)
16:50:35.040     [Warn]  [0x7f352d6a88] cURL [TEXT] h2 header: Authorization:Bearer bla
16:50:35.040    [Debug]  [0x7f352d6a88] cURL [TEXT] STATE: DO => DO_DONE handle 0x7f352d6a88; line 1695 (connection #3892)
16:50:35.040     [Warn]  [0x7f352d6a88] cURL [HEADER_OUT] GET /api/traffic/v2/tile/134766960/current HTTP/2
16:50:35.040    [Debug]  [0x7f352d6a88] cURL [TEXT] STATE: DO_DONE => WAITPERFORM handle 0x7f352d6a88; line 1822 (connection #3892)
16:50:35.044    [Debug]  [0x7f352d6a88] cURL [TEXT] STATE: WAITPERFORM => PERFORM handle 0x7f352d6a88; line 1837 (connection #3892)
16:50:35.044    [Debug]  [0x7f34657d88] cURL [TEXT] STATE: INIT => CONNECT handle 0x7f34657d88; line 1429 (connection #-5000)
16:50:35.044    [Debug]  [0x7f34657d88] cURL [TEXT] Found bundle for host blubb: 0x7f350ec088 [can multiplex]
16:50:35.044    [Debug]  [0x7f34657d88] cURL [TEXT] MAX_CONCURRENT_STREAMS reached, skip (128)
16:50:35.044    [Debug]  [0x7f34657d88] cURL [TEXT] Conn: 3892 (0x7f34af6ca8) Receive pipe weight: (-1/0), penalized: FALSE
16:50:35.044    [Debug]  [0x7f34657d88] cURL [TEXT] Multiplexed connection found!
16:50:35.044    [Debug]  [0x7f34657d88] cURL [TEXT] Found connection 3892, with requests in the pipe (69)
16:50:35.044    [Debug]  [0x7f34657d88] cURL [TEXT] Re-using existing connection! (#3892) with host blubb
16:50:35.044    [Debug]  [0x7f34657d88] cURL [TEXT] STATE: CONNECT => DO handle 0x7f34657d88; line 1474 (connection #3892)
16:50:35.044    [Debug]  [0x7f34657d88] cURL [TEXT] http2_send len=1074
16:50:35.044    [Debug]  [0x7f34657d88] cURL [TEXT] h2 header: :method:GET
16:50:35.044    [Debug]  [0x7f34657d88] cURL [TEXT] h2 header: :path:/api/traffic/v2/tile/134766962/current
16:50:35.044    [Debug]  [0x7f34657d88] cURL [TEXT] h2 header: :scheme:https
16:50:35.044    [Debug]  [0x7f34657d88] cURL [TEXT] h2 header: :authority:blubb
16:50:35.044    [Debug]  [0x7f34657d88] cURL [TEXT] h2 header: Accept-Encoding:deflate, gzip
16:50:35.045    [Debug]  [0x7f34657d88] cURL [TEXT] h2 header: accept:application/x-protobuf
16:50:35.045    [Debug]  [0x7f34657d88] cURL [TEXT] h2 header: content-type:application/x-protobuf
16:50:35.045    [Debug]  [0x7f34657d88] cURL [TEXT] Using Stream ID: e7 (easy handle 0x7f34657d88)
16:50:35.045     [Warn]  [0x7f34657d88] cURL [TEXT] h2 header: Authorization:Bearer bla
16:50:35.045     [Warn]  [0x7f34657d88] cURL [HEADER_OUT] GET /api/traffic/v2/tile/134766962/current HTTP/2
16:50:35.045    [Debug]  [0x7f34657d88] cURL [TEXT] STATE: DO => DO_DONE handle 0x7f34657d88; line 1695 (connection #3892)
16:50:35.045    [Debug]  [0x7f34657d88] cURL [TEXT] STATE: DO_DONE => WAITPERFORM handle 0x7f34657d88; line 1822 (connection #3892)
16:50:35.045    [Debug]  [0x7f34657d88] cURL [TEXT] STATE: WAITPERFORM => PERFORM handle 0x7f34657d88; line 1837 (connection #3892)
16:50:35.046    [Debug]  [0x7f34c84ab8] cURL [TEXT] Expire cleared
16:50:35.048    [Debug]  [0x7f34c69578] cURL [TEXT] Curl_readwrite: forcibly told to drain data
16:50:35.048    [Debug]  [0x7f34c69578] cURL [TEXT] http2_recv: easy 0x7f34c69578 (stream 411)

```

**Comment by ThomasObenaus** [comment]
Next time again a segfault at Curl_Debug as in my first post.
Again the easy_handle seams to be destroyed.
``` 
#0  __GI__IO_fwrite (buf=..., size=..., count=..., fp=...) at iofwrite.c:37
#1  0x0000007f98d0d484 in Curl_debug (data=..., type=..., ptr=..., size=...) at sendf.c:819
#2  0x0000007f98d0d56c in Curl_infof (data=..., fmt=...) at sendf.c:243
#3  0x0000007f98d3cafc in on_frame_not_send (session=..., frame=..., lib_error_code=..., userp=...) at http2.c:813
#4  0x0000007f98cd2380 in nghttp2_session_mem_send_internal (session=..., data_ptr=..., fast_cb=...) at /home/thomaso/work/projects/bob_c_ares/dev/src/libs/ext-nghttp2/1/workspace/nghttp2-1.17.0/lib/nghttp2_session.c:2914
#5  0x0000007f98cd2c04 in nghttp2_session_send (session=...) at /home/thomaso/work/projects/bob_c_ares/dev/src/libs/ext-nghttp2/1/workspace/nghttp2-1.17.0/lib/nghttp2_session.c:3215
#6  0x0000007f98d3d39c in h2_session_send (data=..., h2=...) at http2.c:1478
#7  0x0000007f98d3ee34 in http2_recv (conn=..., sockindex=..., mem=..., len=..., err=...) at http2.c:1658
#8  0x0000007f98d0d2e0 in Curl_read (conn=..., sockfd=..., buf=..., sizerequested=..., n=...) at sendf.c:749
#9  0x0000007f98d1ddb0 in readwrite_data (comeback=..., done=..., didwhat=..., k=..., conn=..., data=...) at transfer.c:482
#10 Curl_readwrite (conn=..., data=..., done=..., comeback=...) at transfer.c:1125
#11 0x0000007f98d28144 in multi_runsingle (multi=..., now=..., data=...) at multi.c:1914
#12 0x0000007f98d28b8c in curl_multi_perform (multi=..., running_handles=...) at multi.c:2181
#13 0x00000000017e6ff0 in tsd::nav::cloud::comm::client::HTTPRequestHandler::main() ()
#14 0x00000000027a3ecc in tsd::nav::cloud::comm::client::HTTPRequestHandler::run() ()
#15 0x0000007f97097510 in tsd::common::system::Thread::startImpl (arg=...) at /home/thomaso/work/projects/bob_c_ares/dev/src/system/tsd-common/1/workspace/tsd.common/src/tsd/common/system/Thread.cpp:310
#16 0x0000007f9709b040 in tsd::common::system::os::thread::pseudoStartThread (arg=...) at /home/thomaso/work/projects/bob_c_ares/dev/src/system/tsd-common/1/workspace/tsd.common/src/tsd/common/system/posix/POSIXThread.cpp:367
#17 0x0000007f96ff1030 in start_thread (arg=...) at pthread_create.c:331
#18 0x0000007f96ce24c0 in thread_start () at ../sysdeps/unix/sysv/linux/aarch64/clone.S:89
(gdb) p data
Attempt to use a type name as an expression
(gdb) f 1
#1  0x0000007f98d0d484 in Curl_debug (data=..., type=..., ptr=..., size=...) at sendf.c:819
819	      fwrite(s_infotype[type], 2, 1, data->set.err);
(gdb) p data->magic 
$1 = 0x21114638
(gdb) p data
$2 = (struct Curl_easy *) 0x7f2110f368
```

**Comment by bagder** [comment]
You're not going to find nor fix the bug this way.

I think you need to either dig in and debug this issue when you get the crashes, or you should start working on writing up a separate app that uses libcurl the same way your actual app does, that can reproduce the problem so that you can share that with us and we can help out with the debugging.

**Comment by ThomasObenaus** [comment]
Yes you are right. 
#### 7.54.0 vs 7.60.0
The odd thing is that with the newest libcurl version (from master) the crash it seams to 'randomly' on different locations. In contrast to the problem in 7.54.0 for me this looks like a heap-corruption.

#### Reproduction sample
To reproduce I had with 7.54.0 is really hard. The setup is kind of special. I already tried a lot to strip it down to a small sample, without success. So there I see no option.

For reproducing the 7.60.0 problem I can try it again.

**Comment by ThomasObenaus** [comment]
But anyway can you give me a hint where I can look at (in libcurl) to find the cause of the crashes. I have not enough knowledge about how curl handles the connections and shares them between easy handles.

**Comment by bagder** [comment]
I asked [questions previously](https://github.com/curl/curl/issues/2688#issuecomment-400829901) whose answers could help.

You also need to describe with more details what you do with libcurl and what h2 streams that are ongoing (or not) when the problem occurs. Possibly even what exact h2 frame that triggers the problem.

**Comment by ThomasObenaus** [comment]
Ok sorry. To answer your questions:
1. Sadly running it with valgrind is not possible since its a too big application.
2. I'll try to find the potential h2 frame that causes the problem.
3. I'll prepare a bit of pseudo-code to give you an idea how I (maybe wrongly) use curl.

**Comment by bagder** [comment]
When `data->magic` is wrong, is the rest of the struct also messed up? Since the pointer seems to still be the right one, that would indicate that the memory was either freed or overwritten, right?

Given the log, it seems like that handle *should* still be alive and have an ongoing associated transfer?

**Comment by ThomasObenaus** [comment]
I'll do another run to check this..

**Comment by ThomasObenaus** [comment]
### How we use curl
```
class Request
{
private:
  CURL *m_CurlEasyHandle = nullptr;
  std::function<void(std::shared_ptr<Request> request)> m_OnComplete;
};

void init()
{
  MutexGuard lock(m_Mutex);
  m_CurlMultiHandle = curl_multi_init();

  // enable http2 multiplexing (which is the whole purpose of this code)
  curl_multi_setopt(m_CurlMultiHandle, CURLMOPT_PIPELINING, CURLPIPE_MULTIPLEX);

  // enable http2 push
  curl_multi_setopt(m_CurlMultiHandle, CURLMOPT_PUSHFUNCTION, &pushCallback);
  curl_multi_setopt(m_CurlMultiHandle, CURLMOPT_PUSHDATA, this);

  // start performing
  curl_multi_perform(m_CurlMultiHandle, &m_StillRunning);
}

void performRequests(std::queue<std::shared_ptr<Request>> requests)
{
  while (!requests.empty())
  {
    performRequest(requests.front())
        requests.pop();
  }
}

void performRequest(const std::shared_ptr<Request> &request)
{
  if (!request)
    return;

  request->m_CurlEasyHandle = curl_easy_init();
  if (!request->m_CurlEasyHandle)
    return;

  // http request headers
  curl_slist *header = nullptr;
  header = addAccessTokenToHeader(header);
  header = addCommonHeaders(header);

  auto &refHeader = request->m_Request.getHeader();
  if (!refHeader.empty())
  {
    for (auto &refElement : refHeader)
    {
      auto &refHeaderField = refElement.second;
      header = curl_slist_append(header, refHeaderField.toString().c_str());
    }
  }

  // add the header
  curl_easy_setopt(request->m_CurlEasyHandle, CURLOPT_HTTPHEADER, headers);
  curl_easy_setopt(request->m_CurlEasyHandle, CURLOPT_TIMEOUT, 10);
  curl_easy_setopt(request->m_CurlEasyHandle, CURLOPT_DNS_CACHE_TIMEOUT, 60);
  curl_easy_setopt(request->m_CurlEasyHandle, CURLOPT_URL, "https://blubb.com/bla/blubb");
  curl_easy_setopt(request->m_CurlEasyHandle, CURLOPT_FOLLOWLOCATION, 1L);  // follow http 30x redirects
  curl_easy_setopt(request->m_CurlEasyHandle, CURLOPT_ACCEPT_ENCODING, ""); // accept all compression algorithms we know
  curl_easy_setopt(request->m_CurlEasyHandle, CURLOPT_PRIVATE, header);
  curl_easy_setopt(request->m_CurlEasyHandle, CURLOPT_ERRORBUFFER, m_errBuffer); // provide a buffer to store errors in
  m_errBuffer[0] = 0;                                                            // set the error buffer as empty before performing a request
  curl_easy_setopt(request->m_CurlEasyHandle, CURLOPT_NOSIGNAL, 1L);
  // always use http2 without sending an http1.x request first
  curl_easy_setopt(request->m_CurlEasyHandle, CURLOPT_HTTP_VERSION, CURL_HTTP_VERSION_2_PRIOR_KNOWLEDGE);
  curl_easy_setopt(request->m_CurlEasyHandle, CURLOPT_SSL_VERIFYHOST, 0L);
  curl_easy_setopt(request->m_CurlEasyHandle, CURLOPT_SSL_VERIFYPEER, 0L);
  curl_easy_setopt(request->m_CurlEasyHandle, CURLOPT_VERBOSE, 1L);
  curl_easy_setopt(request->m_CurlEasyHandle, CURLOPT_DEBUGFUNCTION, curlDebugCallback);
  curl_easy_setopt(request->m_CurlEasyHandle, CURLOPT_DEBUGDATA, &m_debugLoggingCurlData);
  curl_easy_setopt(request->m_CurlEasyHandle, CURLOPT_WRITEFUNCTION, &writeMessage);
  curl_easy_setopt(request->m_CurlEasyHandle, CURLOPT_WRITEDATA, request.get());
  curl_easy_setopt(request->m_CurlEasyHandle, CURLOPT_HEADERFUNCTION, &readHeader);
  curl_easy_setopt(request->m_CurlEasyHandle, CURLOPT_HEADERDATA, request.get());

  if (request->httpMethod == HttpMethod::GET)
  {
    // no request data, send as GET request
    curl_easy_setopt(request->m_CurlEasyHandle, CURLOPT_HTTPGET, 1);
  }
  else if (request->httpMethod == HttpMethod::POST)
  {
    // send request as POST request
    curl_easy_setopt(request->m_CurlEasyHandle, CURLOPT_POST, 1);
    curl_easy_setopt(request->m_CurlEasyHandle, CURLOPT_POSTFIELDS, request->m_Request.data().data());
    curl_easy_setopt(request->m_CurlEasyHandle, CURLOPT_POSTFIELDSIZE, request->m_Request.data().size());
  }

  curl_multi_add_handle(m_CurlMultiHandle, request->m_CurlEasyHandle);
  m_Requests[request->m_CurlEasyHandle] = request;
}

std::shared_ptr<Request>
httpRequestFind(const CURL *const easy)
{
  auto requestIt = m_Requests.find(easy);
  if (requestIt == m_Requests.end())
  {
    // should never happen
    throw Exception("implementation error: could not find request in list");
  }
  return requestIt->second;
}

void httpRequestDone(const std::shared_ptr<Request> request, const CURLcode &curlCode)
{
  if (request)
  {
    if (curlCode == CURLE_OK)
    {
      long status = 0;
      curl_easy_getinfo(request->m_CurlEasyHandle, CURLINFO_RESPONSE_CODE, &status);
      request->m_Status = HTTPStatusCode(status);
    }
    else
    {
      std::string curlErr(curl_easy_strerror(curlCode));

      std::stringstream ss;
      ss << std::hex << "[" << request->m_CurlEasyHandle << "]"
         << "CurlError: " << curlErr << " (" << curlCode << ")" << &std::endl;
      LOG_ERROR << ss.str();

      auto len = strlen(m_errBuffer);
      if (len)
      {
        std::stringstream ss;
        ss << std::hex << "[" << request->m_CurlEasyHandle << "]"
           << "Error info: " << std::string(m_errBuffer) << &std::endl;
        LOG_ERROR << ss.str();
      }
    }

    // call the requester and put the response to him
    request->m_OnComplete(request);

    // get header and delete
    curl_slist *header = nullptr;
    curl_easy_getinfo(request->m_CurlEasyHandle, CURLINFO_PRIVATE, &header);
    curl_slist_free_all(header);

    const CURLMcode cCode = curl_multi_remove_handle(m_CurlMultiHandle, request->m_CurlEasyHandle);
    auto countErased = m_Requests.erase(request->m_CurlEasyHandle);

    curl_easy_cleanup(request->m_CurlEasyHandle);
    request->m_CurlEasyHandle = nullptr;
  }
}

void mainLoop()
{
  curl_global_init(CURL_GLOBAL_ALL);
  init();
  do
  {
    if (getPendingRequests() < MAX_PENDING_REQUESTS)
      performRequests(popWaitingRequests(MAX_PENDING_REQUESTS));

    fd_set fdRead;
    fd_set fdWrite;
    fd_set fdExcep;

    FD_ZERO(&fdRead);
    FD_ZERO(&fdWrite);
    FD_ZERO(&fdExcep);

    // set a suitable timeout to play around with
    long curlTimeout = -1;
    curl_multi_timeout(m_CurlMultiHandle, &curlTimeout);

    const long maxTimeoutMs = 1000;
    timeval timeout = curlTimeoutToTimeval(curlTimeout, maxTimeoutMs);

    // get file descriptors from the transfers
    int maxfd = -1;
    {
      CURLMcode code = curl_multi_fdset(m_CurlMultiHandle, &fdRead, &fdWrite, &fdExcep, &maxfd);
      if (CURLM_OK != code)
      {
        throw ExceptionCurlMulti(code);
      }
    }

    int rc = 0;
    if (-1 == maxfd)
      sleep(100);
    else
      rc = select(maxfd + 1, &fdRead, &fdWrite, &fdExcep, &timeout);

    if (-1 != rc)
    {
      // timeout or readable/writable sockets
      CURLMcode cCMultiPerform = curl_multi_perform(m_CurlMultiHandle, &m_StillRunning);

      // check curl for messages
      CURLMsg *msg = nullptr;
      do
      {
        // get messages
        int numMessages = 0;
        msg = curl_multi_info_read(m_CurlMultiHandle, &numMessages);

        // check if a request is done
        if (msg && (CURLMSG_DONE == msg->msg))
        {
          const CURLcode code = msg->data.result;
          auto request = httpRequestFind(msg->easy_handle);
          httpRequestDone(request, code);
        }
      } while (msg);
    }

  } while (!m_Shutdown);

  curl_multi_cleanup(m_CurlMultiHandle);
}
```

**Comment by ThomasObenaus** [comment]
### In short
```
easy_handle = curl_easy_init();
configure(easy_handle);
curl_multi_timeout(m_CurlMultiHandle, &curlTimeout);
CURLMcode code = curl_multi_fdset(m_CurlMultiHandle, &fdRead, &fdWrite, &fdExcep, &maxfd);
int rc = select(maxfd + 1, &fdRead, &fdWrite, &fdExcep, &timeout);
CURLMcode cCMultiPerform = curl_multi_perform(m_CurlMultiHandle, &m_StillRunning);
// check curl for messages
CURLMsg *msg = nullptr;
for all (easy_handles given back by curl)
{
 int numMessages = 0;
 msg = curl_multi_info_read(m_CurlMultiHandle, &numMessages); 
 // check if a request is done
 if (msg && (CURLMSG_DONE == msg->msg))
 {
   const CURLcode code = msg->data.result;
   auto request = httpRequestFind(msg->easy_handle);
   curl_slist *header = nullptr;
   curl_easy_getinfo(request->m_CurlEasyHandle, CURLINFO_PRIVATE, &header);
   curl_slist_free_all(header);

    const CURLMcode cCode = curl_multi_remove_handle(m_CurlMultiHandle, request->m_CurlEasyHandle);
    auto countErased = m_Requests.erase(request->m_CurlEasyHandle);
    curl_easy_cleanup(request->m_CurlEasyHandle);
    request->m_CurlEasyHandle = nullptr;
  }
} while (msg);
```

**Comment by ThomasObenaus** [comment]
This looks like the problem:

```on_stream_close``` is called on the easy_handle which was freed with ```curl_easy_cleanup``` shortly before

```
13:39:52.853     [Warn] [user] nav: CURL.H2 [easy_handle=0x7f38a92558,magic=0xc0dedbad,tid=3233] http2_recv: easy 0x7f38a92558 (stream 139)
13:40:13.149     [Warn] [user] nav: HTTPR: httpRequestDone->curl_easy_cleanup (curl_easy=0x7f382a49a8,req=0x7f3007a690,tid=3233)...
13:40:13.149     [Warn] [user] nav: HTTPR: httpRequestDone->curl_easy_cleanup (curl_easy=0x7f382a49a8,req=0x7f3007a690,tid=3233)...done
13:40:21.831     [Warn] [user] nav: CURL.H2 [easy_handle=0x7f382a49a8,magic=(nil),tid=3233] on_stream_close(), STREAM_CLOSED (err 5), stream 139
```

But how could that happen? We access the easy_handle only from one thread and we only pass it to our httpRequestDone method where ```curl_easy_cleanup``` is called **after** we got the handle back from curl with:
```
 CURLMsg *msg = curl_multi_info_read(m_CurlMultiHandle, &numMessages); 
 // check if a request is done
 if (msg && (CURLMSG_DONE == msg->msg))
   httpRequestDone(); 
```

**Comment by ThomasObenaus** [comment]
### Full life-cycle of this easy_handle
```
13:40:02.988     [Warn] [user] nav: HTTPR: performRequest->curl_easy_init (curl_easy=0x7f382a49a8,req=0x7f3007a690,tid=3233)...done
13:40:02.988     [Warn] [user] nav: HTTPR: performRequest->curl_multi_add_handle (curl_easy=0x7f382a49a8,req=0x7f3007a690,tid=3233)...
13:40:02.988     [Warn] [user] nav: HTTPR: performRequest->curl_multi_add_handle (curl_easy=0x7f382a49a8,req=0x7f3007a690,tid=3233,cCmultiAdd=0)...done
13:40:03.079     [Warn] [user] nav: [0x7f382a49a8] cURL [TEXT] STATE: INIT => CONNECT handle 0x7f382a49a8; line 1429 (connection #-5000)
13:40:03.079     [Warn] [user] nav: [0x7f382a49a8] cURL [TEXT] Found bundle for host blubb: 0x7f3851d858 [can multiplex]
13:40:03.080     [Warn] [user] nav: [0x7f382a49a8] cURL [TEXT] Conn: 8418 (0x7f38afe1d8) Receive pipe weight: (-1/0), penalized: FALSE
13:40:03.080     [Warn] [user] nav: [0x7f382a49a8] cURL [TEXT] Multiplexed connection found!
13:40:03.080     [Warn] [user] nav: [0x7f382a49a8] cURL [TEXT] Found connection 8418, with requests in the pipe (55)
13:40:03.080     [Warn] [user] nav: [0x7f382a49a8] cURL [TEXT] Re-using existing connection! (#8418) with host blubb
13:40:03.080     [Warn] [user] nav: [0x7f382a49a8] cURL [TEXT] STATE: CONNECT => DO handle 0x7f382a49a8; line 1474 (connection #8418)
13:40:03.080     [Warn] [user] nav: CURL.H2 [easy_handle=0x7f382a49a8,magic=0xc0dedbad,tid=3233] http2_send len=1074
13:40:03.080     [Warn] [user] nav: CURL.H2 [easy_handle=0x7f382a49a8,magic=0xc0dedbad,tid=3233] h2 header: :method:GET
13:40:03.080     [Warn] [user] nav: CURL.H2 [easy_handle=0x7f382a49a8,magic=0xc0dedbad,tid=3233] h2 header: :path:/api/traffic/v2/tile/134766962/current
13:40:03.080     [Warn] [user] nav: CURL.H2 [easy_handle=0x7f382a49a8,magic=0xc0dedbad,tid=3233] h2 header: :scheme:https
13:40:03.080     [Warn] [user] nav: CURL.H2 [easy_handle=0x7f382a49a8,magic=0xc0dedbad,tid=3233] h2 header: :authority:blubb
13:40:03.080     [Warn] [user] nav: CURL.H2 [easy_handle=0x7f382a49a8,magic=0xc0dedbad,tid=3233] h2 header: Accept-Encoding:deflate, gzip
13:40:03.080     [Warn] [user] nav: CURL.H2 [easy_handle=0x7f382a49a8,magic=0xc0dedbad,tid=3233] h2 header: accept:application/x-protobuf
13:40:03.080     [Warn] [user] nav: CURL.H2 [easy_handle=0x7f382a49a8,magic=0xc0dedbad,tid=3233] h2 header: content-type:application/x-protobuf
13:40:03.080     [Warn] [user] nav: CURL.NGHTTP2 nghttp2_submit_request session=0x7f39165db0,stream_user_data=0x7f382a49a8,tid=3233
13:40:03.080     [Warn] [user] nav: CURL.NGHTTP2 submit_headers_shared_nva session=0x7f39165db0,stream_user_data=0x7f382a49a8,tid=3233
13:40:03.080     [Warn] [user] nav: CURL.NGHTTP2 submit_headers_shared session=0x7f39165db0,stream_user_data=0x7f382a49a8,tid=3233
13:40:03.080     [Warn] [user] nav: [0x7f382a49a8] cURL [TEXT] Using Stream ID: 8b (easy handle 0x7f382a49a8)
13:40:03.080     [Warn] [user] nav: [0x7f382a49a8] cURL [HEADER_OUT] GET /api/traffic/v2/tile/134766962/current HTTP/2
13:40:03.080     [Warn] [user] nav: [0x7f382a49a8] cURL [TEXT] STATE: DO => DO_DONE handle 0x7f382a49a8; line 1695 (connection #8418)
13:40:03.080     [Warn] [user] nav: [0x7f382a49a8] cURL [TEXT] STATE: DO_DONE => WAITPERFORM handle 0x7f382a49a8; line 1822 (connection #8418)
13:40:03.080     [Warn] [user] nav: [0x7f382a49a8] cURL [TEXT] STATE: WAITPERFORM => PERFORM handle 0x7f382a49a8; line 1837 (connection #8418)
13:40:13.082     [Warn] [user] nav: [0x7f382a49a8] cURL [TEXT] Operation timed out after 10000 milliseconds with 0 bytes received
13:40:13.082     [Warn] [user] nav: [0x7f382a49a8] cURL [TEXT] Kill stream: Disconnected with pending data
13:40:13.082     [Warn] [user] nav: [0x7f382a49a8] cURL [TEXT] multi_done
13:40:13.082     [Warn] [user] nav: [0x7f382a49a8] cURL [TEXT] Connection still in use 0/44, no more multi_done now!
13:40:13.148     [Info] [HTTPRequestHandler] [0x7f382a49a8] Response received from curl.
13:40:13.148    [Error] [HTTPRequestHandler] [0x7f382a49a8]CurlError: Timeout was reached (1c)
13:40:13.148    [Error] [HTTPRequestHandler] [0x7f382a49a8]Error info: Failed sending HTTP POST request
13:40:13.148     [Info] [HTTPRequestHandler] [0x7f382a49a8] httpRequestDone
13:40:13.149     [Warn] [user] nav: HTTPR: main->curl_multi_info_read (curl_multi=0x1e28c988,numMessages=8,tid=3233,msg=0x7f382a4a10easy_handle=0x7f382a49a8msg.msg=1)...done
13:40:13.149     [Warn] [user] nav: HTTPR: main->httpRequestFind (easy_handle=0x7f382a49a8)...
13:40:13.149     [Warn] [user] nav: HTTPR: main->httpRequestFind (easy_handle=0x7f382a49a8,request=0x7f3007a690)...done
13:40:13.149     [Warn] [user] nav: HTTPR: httpRequestDone (curl_easy=0x7f382a49a8,req=0x7f3007a690,tid=3233)...
13:40:13.149     [Warn] [user] nav: HTTPR: httpRequestDone->curl_multi_remove_handle (curl_easy=0x7f382a49a8,req=0x7f3007a690,tid=3233)...
13:40:13.149     [Warn] [user] nav: HTTPR: httpRequestDone->curl_multi_remove_handle (curl_easy=0x7f382a49a8,req=0x7f3007a690,cCode=0,tid=3233)...done
13:40:13.149     [Warn] [user] nav: HTTPR: m_Requests.erase (easy_handle=0x7f382a49a8,request=0x7f3007a690,cCode=0)
13:40:13.149     [Warn] [user] nav: HTTPR: httpRequestDone->curl_easy_cleanup (curl_easy=0x7f382a49a8,req=0x7f3007a690,tid=3233)...
13:40:13.149     [Warn] [user] nav: HTTPR: httpRequestDone->curl_easy_cleanup (curl_easy=0x7f382a49a8,req=0x7f3007a690,tid=3233)...done
13:40:13.149     [Warn] [user] nav: HTTPR: httpRequestDone (curl_easy=0x7f382a49a8,req=0x7f3007a690,tid=3233)...done
13:40:21.831     [Warn] [user] nav: CURL.NGHTTP2 nghttp2_session_open_stream 3 session=0x7f39165db0,stream_user_data=0x7f382a49a8,tid=3233
13:40:21.831     [Warn] [user] nav: CURL.NGHTT2 nghttp2_session_open_stream 3434 session=0x7f39165db0,stream_user_data=0x7f382a49a8,tid=3233
13:40:21.831     [Warn] [user] nav: CURL.NGHTTP2 ->nghttp2_stream_init 5545 session=0x7f39165db0,stream_user_data=0x7f382a49a8,tid=3233
13:40:21.831     [Warn] [user] nav: CURL.NGHTTP2 nghttp2_stream_init stream_user_data=0x7f382a49a8,tid=3233
13:40:21.831     [Warn] [user] nav: CURL.H2 [easy_handle=0x7f382a49a8,magic=(nil),tid=3233] on_frame_not_send() was called, lib_error_code = -511
13:40:21.831     [Warn] [user] nav: CURL.H2 [easy_handle=0x7f382a49a8,magic=(nil),tid=3233] on_stream_close(), STREAM_CLOSED (err 5), stream 139
```

**Comment by bagder** [comment]
> But how could that happen?

Well, there's a bug somewhere in here...

When `on_stream_close` it seems the call to `nghttp2_session_get_stream_user_data` within that function returns the old handle for a handle that is already gone and killed?

First: how come there is frames for that stream arriving after it was deemed done? Probably because the transfer was timed out or otherwise prematurely ended from curl's side.

But more importantly: when a transfer is ended, `Curl_http_done` should be called, which in turn calls `Curl_http2_done`. And in there it calls, or *should call*, `nghttp2_session_set_stream_user_data()` to clear the association from the stream-id to this now ending transfer (easy handle).

So, the incoming frame on an already completed transfer should *not* be associated with the old easy handle when things work out. Can you figure out why this logic doesn't work for your scenario?

**Comment by ThomasObenaus** [comment]
I'll check this on Monday. Sadly currently I don't have access to the setup.

**Comment by jeroen** [comment]
@ThomasObenaus does your application use `CURLMOPT_MAX_HOST_CONNECTIONS`? I found that this seems to trigger problems in my case.

**Comment by ThomasObenaus** [comment]
Hey @jeroen I tried it with:
```
curl_multi_setopt( m_CurlMultiHandle, CURLMOPT_PIPELINING, CURLPIPE_MULTIPLEX );
curl_multi_setopt( m_CurlMultiHandle, CURLMOPT_MAX_HOST_CONNECTIONS, 2L );
curl_multi_setopt( m_CurlMultiHandle, CURLMOPT_MAX_TOTAL_CONNECTIONS, 2L );
curl_multi_setopt( m_CurlMultiHandle, CURLMOPT_MAXCONNECTS, 10L );
```

**Comment by bagder** [comment]
@ThomasObenaus, in my work with #2674 based on @jeroen's example code, I also experience what looks almost exactly like your case. I've not yet figured it out, but at least there's hope that given more time I might have fixes pending for both these issues.

Part of my recent debugging session was streamed over twitch and can be [seen again](https://www.twitch.tv/videos/280754395) (two hours of heavy gdb usage!). Unfortunately I had to pause it before I could call victory. I have fixed a few issues along the way though, even if it still isn't done.

**Comment by ReneReusner** [comment]
Okay cause the old issue was closed:


Hello. I managed to get some hours at work to look in this problem again. What i had found: On a curl time out in multi.c (~1400)       timeout_ms = Curl_timeleft(data, &now,                                  (data->mstate <= CURLM_STATE_WAITDO)?                                  TRUE:FALSE);        if(timeout_ms < 0) {         /* Handle timed out */ ............          /* Force connection closed if the connection has indeed been used */         if(data->mstate > CURLM_STATE_DO) {           streamclose(data->easy_conn, "Disconnected with pending data");           stream_error = TRUE;         }         result = CURLE_OPERATION_TIMEDOUT;         (void)multi_done(&data->easy_conn, result, TRUE);         /* Skip the statemachine and go directly to error handling section. */         goto statemachine_end; The multi_done here static CURLcode multi_done(struct connectdata **connp,                           CURLcode status,  /* an error if this is called                                                after an error was detected */                           bool premature) needs to clear the user data in the associated stream. If it don't to  it, then sometimes when nghttp2 iterate over all stream in a session,  sometimes it will find a "stale" user_data and may call a callback with a  deleted Curl_easy -> BOOM. I added it in void Curl_http2_done(struct connectdata *conn, bool premature) { Please also take a look in multi.c ~ 1350 static CURLMcode multi_runsingle(struct Curl_multi *multi,                                  struct timeval now,                                  struct Curl_easy *data) ...       data->state.pipe_broke = FALSE;       data->easy_conn = NULL;       /* My comment for a debug assertion          clear http2 session infos, the http2 session is already gone,          but the assert in http2.c:1979          assert(!(conn->data->tmp1) \|\| (conn->data->tmp1 == h2)); & co          can still bite us, if we get a new connection to continue */ On a broken pipe, Cause i'm not familiar with curl, i don't know if  the associated connection can be still used? If yes (maybe its placed in  the connection cache), we need to remove the stream_userdata here too. One another thing. Its not a error but can break on 64bit machines a backtrace in gdb: sendf.c ~ line 865:   if(data->set.fdebug) {     return curlDebugCallback(data, type, ptr, size,  data->set.debugdata);     /*return (*data->set.fdebug)(data, type, ptr, size, For example on a stale Curl_easy, set.fdebug points to a invalid  address, when the cpu jump to it, it crashes yes. BUT (cause of some  brain damaged ABI designers, redzone, framepointer optional) the gdb  needs a "valid" PC to get the unwind information. If its not in a known  region -> no backtrace. Then you need a walk_stack macro. Callbacks  for debugging on 64bit machines are not always a good idea. A callback which point to a invalid address (for example from a old stale Curl_easy) can result in a broken backtrace.
--

**Comment by ReneReusner** [comment]
So after putting in a lot of assertion, i think i'm on the 1/3 way to the issue (next 2.5 weeks holiday, so i'm not able to work on it).

I hope I can write it understandably.

One assumption in libcurl http2 is: a session consists of streams. A curl_easy is associated to a stream in a session. To "remove" a stream in Curl_http2_done from a session is by removing the user_data from the session.

```c
void Curl_http2_done(struct connectdata *conn, bool premature)
.....
nghttp2_session_set_stream_user_data(httpc->h2, http->stream_id, 0);
```

The assumption is: still maybe "in the stream" but it won't be referenced, cause we don't have a userdata to it -> callbacks find no Curl_easy to it by the given streamid.

After instrumenting the callbacks, some assertions and even the stream hash map in nghttp2 i can say:

This assumption is wrong. When a new stream is created its not placed in

```c
(gdb) p *conn->proto.httpc.h2
$16 = {
  streams = {
    table = 0x7f352cff00, 
    mem = 0x7f352c4b90, 
    size = 0x1, 
    tablelen = 0x100
  }, 
```
Its still queued elsewhere.

A stream is created in:

```c
static ssize_t http2_send(struct connectdata *conn, int sockindex,
                          const void *mem, size_t len, CURLcode *err)
.....
    stream_id = nghttp2_submit_request(h2, &pri_spec, nva, nheader,
                                       &data_prd, conn->data);
```

After this its still NOT in the map, its queued elsewhere.

So when nghttp2 detect it can do some sending of this stream, it will dequeue it from this place and put this in the streamid hash map. When this code path is taken, nghttp2 will call the callback:

```c
static int before_frame_send(nghttp2_session *session,
                             const nghttp2_frame *frame,
                             void *userp)
```

After this, its in the map.

So what is happening in the error case. When we "detect" a timeout on a stream, i call it B, which still did not passed the

```c
static int before_frame_send(nghttp2_session *session,
```
state, then

```c
static CURLMcode multi_runsingle(struct Curl_multi *multi,
                                 struct timeval now,
                                 struct Curl_easy *data)
```

will call:

```c
        (void)multi_done(&data->easy_conn, result, TRUE);
```
which will call

```c
void Curl_http2_done(struct connectdata *conn, bool premature)
```
However the

```c
nghttp2_session_set_stream_user_data(httpc->h2, http->stream_id, 0);
```
from stream B will fail, cause its still not in the map. After this the associated Curl_easy *data is "completed" with a timeout and will be deleted from our HTTPRequestHandler.

BUT:
When handling other streams in this session, nghttp2 will detect "oh it can send something from a new stream B".

gdb backtrace:
```c
(gdb) bt
#0  0x0000007facd3b1e8 in __GI_raise (sig=...) at ../sysdeps/unix/sysv/linux/raise.c:54
#1  0x0000007facd3c5dc in __GI_abort () at abort.c:89
#2  0x0000007facd3494c in __assert_fail_base (fmt=..., assertion=..., file=..., line=..., function=...) at assert.c:92
#3  0x0000007facd349e4 in __GI___assert_fail (assertion=..., file=..., line=..., function=...) at assert.c:101
#4  0x0000007faee28bc8 in on_frame_not_send (session=..., frame=..., lib_error_code=..., userp=...) at http2.c:769
#5  0x0000007faedbc460 in nghttp2_session_mem_send_internal (session=..., data_ptr=..., fast_cb=...)
    at /home/rene/src/cares/dev/src/libs/ext-nghttp2/1/workspace/nghttp2-1.17.0/lib/nghttp2_session.c:2920
#6  0x0000007faedbcce4 in nghttp2_session_send (session=...)
    at /home/rene/src/cares/dev/src/libs/ext-nghttp2/1/workspace/nghttp2-1.17.0/lib/nghttp2_session.c:3221
#7  0x0000007faee2a010 in http2_send (conn=..., sockindex=..., mem=..., len=..., err=...) at http2.c:2084
#8  0x0000007faedf691c in Curl_write (conn=..., sockfd=..., mem=..., len=..., written=...) at sendf.c:341
#9  0x0000007faedf33ec in Curl_add_buffer_send (in=..., conn=..., bytes_written=..., included_body_bytes=..., 
    socketindex=...) at http.c:1128
#10 0x0000007faedf49f0 in Curl_http (conn=..., done=...) at http.c:2710
#11 0x0000007faee14d10 in multi_do (done=..., connp=...) at multi.c:1266
#12 multi_runsingle (multi=..., now=..., data=...) at multi.c:1685
#13 0x0000007faee1563c in curl_multi_perform (multi=..., running_handles=...) at multi.c:2206
#14 0x00000000017e6c90 in tsd::nav::cloud::comm::client::HTTPRequestHandler::main (this=...)
    at /home/rene/src/cares/dev/src/navigation/tsd-nav-cloud-comm-client/1/workspace/tsd.nav.cloud.comm.client/src/tsd/nav/cloud/comm/client/HTTPRequestHandler.cpp:1271
#15 0x00000000027a4b6c in tsd::nav::cloud::comm::client::HTTPRequestHandler::run (this=...)
    at /home/rene/src/cares/dev/src/navigation/tsd-nav-cloud-comm-client/1/workspace/tsd.nav.cloud.comm.client/src/tsd/nav/cloud/comm/client/HTTPRequestHandler.cpp:902
#16 0x0000007fad184510 in tsd::common::system::Thread::startImpl (arg=...)
    at /home/rene/src/cares/dev/src/system/tsd-common/1/workspace/tsd.common/src/tsd/common/system/Thread.cpp:310
#17 0x0000007fad188040 in tsd::common::system::os::thread::pseudoStartThread (arg=...)
    at /home/rene/src/cares/dev/src/system/tsd-common/1/workspace/tsd.common/src/tsd/common/system/posix/POSIXThread.cpp:367
#18 0x0000007fad0de030 in start_thread (arg=...) at pthread_create.c:331
#19 0x0000007facdcf4c0 in thread_start () at ../sysdeps/unix/sysv/linux/aarch64/clone.S:89
(gdb) f 12
#12 multi_runsingle (multi=..., now=..., data=...) at multi.c:1685
1685            result = multi_do(&data->easy_conn, &dophase_done);
(gdb) p data
$17 = (struct Curl_easy *) 0x7f34af6118        <---- STREAM A
(gdb) f 4
#4  0x0000007faee28bc8 in on_frame_not_send (session=..., frame=..., lib_error_code=..., userp=...) at http2.c:769
769         assert(data_s->easy_conn == conn);
(gdb) p data_s
$18 = (struct Curl_easy *) 0x7f345a4b28         <-- STREAM B
(gdb) 
```

 nghttp2 will place the the stream B in the stream hash map.
We have the the case, we will get a callback for "invalid" stream like:

```c
static int on_frame_send(nghttp2_session *session,
                         const nghttp2_frame *frame,
                         void *userp)
{
  struct Curl_easy *data_s;
  /* (void)userp; */
  struct connectdata *conn = (struct connectdata *)userp;

  data_s = nghttp2_session_get_stream_user_data(session, frame->hd.stream_id);
  printf("XXXXX on_frame_send: data: %p, conn: %p, session: %p, id: %d\n",
         data_s, conn, session, frame->hd.stream_id);
  if(data_s) {
    assert(data_s->easy_conn == conn);
    DEBUGF(infof(data_s, "on_frame_send() was called, length = %zd\n",
                 frame->hd.length));
  }
  return 0;
```

Which will find stream B it will call "infof", which wants to display the associated data_s from stream B.
Cause its the deleted, the callback pointer is overwritten -> jump to nirvana and crash.

```c
nghttp2_session_set_stream_user_data(httpc->h2, http->stream_id, 0);
```
Is just not enough, to "clear" streamid and associated user_data. We need to dequeue it from the nghttp_session.

**Comment by bagder** [comment]
> To "remove" a stream in "Curl_http2_done" from a session is by removing the user_data from the session.

I wouldn't call that to *remove* the stream. It really doesn't. That's supposed to *remove the association* between the stream_id and the easy handle. So if any frames arrive after that point in time with that stream_id, we can see that this is an unmapped stream and we ignore it.

If I understand you correctly. You have identified a case where a h2 stream is started (and user_data is set), curl stops the stream before it gets added to "the map" and tries to clear the user_data - which fails because it isn't yet in the map. After that, a frame from that stream appears and it then (surprise!) provides the pointer as user_data even though we tried to clear it previously.

Was that a roughly correct description?

**Comment by bagder** [comment]
If the set_data that tries to clear the stream before it exists fails, it should return an error code and we should notice that. I've made #2880 just now that adds proper return code checks for the `nghttp2_session_set_stream_user_data()` calls.

**Comment by ReneReusner** [comment]
If I understand you correctly. You have identified a case where a h2 stream is started (and user_data is set), curl stops the stream before it gets added to "the map" and tries to clear the user_data - which fails because it isn't yet in the map. After that, a frame from that stream appears and it then (surprise!) provides the pointer as user_data even though we tried to clear it previously.

Was that a roughly correct description?

YES
nghttp2_session_set_stream_user_data() to zero won't help in this case, cause its added later on.
We now have a maybe deleted Curl_easy* in a callback.

**Comment by bagder** [comment]
> nghttp2_session_set_stream_user_data() to zero won't help in this case

Does it return an error for this case? Do you consider this to be a bug in nghttp2? Do we need to add our own map to detect these completed streams that might give us something we don't want?

**Comment by ReneReusner** [comment]
Yes it return an error.
But the user_data is still added after this -> BOOM.

**Comment by bagder** [comment]
@tatsuhiro-t: can this and is it meant to be able to happen? We call `nghttp2_submit_request()` that adds a stream_id => user_data association (toward the end of the `http2_send` function) but according to @ReneReusner's findings we can end up calling `nghttp2_session_set_stream_user_data()` to clear that association *before* it is actually made, which makes it fail to clear the association and instead we can end up getting that user_data passed anyway if frames from that stream still arrives.

**Comment by ReneReusner** [comment]
> Does it return an error for this case? Do you consider this to be a bug in nghttp2? Do we need to add our own map to detect these completed streams that might give us something we don't want?

After years of debugging alien software, i won't call it a bug, maybe a feature or implementation detail :-).

A own map would help:

> One way is: For the curl connection have a own "set" of active and valid streamids, so we can "detect" streamids with invalid user_data.

It feels like "only a workaround". But i did a lot of "workarounds", so i can live with an another workaround.

**Comment by ReneReusner** [comment]
Back from cinema (mission impossible yay). I hope i can see bugfix after my vacation :-).

But back to topic:

One idea, **the place elsewhere** before the stream is insert in the map is maybe the "data provider (data_prd) interface" of nghttp2, if i remember correctly from my memory. Only when nghttp2 starts sending a stream is insert in the streamid map. The reason why i'm writing this, we need to make sure, there also no remains of a dead stream are left in this interface. In the most cases i would dig deeper, cause i think i'm on the right track, but my 2.5 week vacation starts now. 

Have a nice summer and a lot of ice cream.

**Comment by tatsuhiro-t** [comment]
@bagder Yes, it can happen if we call nghttp2_set_stream_user_data before sending first HEADERS frame which includes HTTP request.  nghttp2_session_set_stream_user_data succeeds only after nghttp2_before_frame_send_callback or nghttp2_on_frame_not_send_callback.
Is Curl_http_done called before sending HTTP request?

**Comment by tatsuhiro-t** [comment]
nghttp2 has a special handling of nghttp2_submit_rst_stream.  If it is used against  a stream which is not opened yet (the case we are concerned about), it cancels the stream and no frame is sent.
nghttp2_on_stream_close_callback is still called in that case, and if http->h2 must be deleted before that, we might access dangling pointer.
A possible resolution to this is use another object which binds a stream and http->h2 and use it as stream user data.

```c
struct data_binder {
    typeof(http->h2) *h2;
};
```

 This object is created when we call nghttp2_submit_request and deleted in nghttp2_on_stream_callback.  Where we call nghttp2_session_set_stream_user_data(s, id, NULL), we can get the object and assign h2 field to NULL.

**Comment by bagder** [comment]
Thanks for these details @tatsuhiro-t !

> it can happen if we call nghttp2_set_stream_user_data before sending first HEADERS frame which includes HTTP request

...

> A possible resolution to this is use another object

OK, but that possible resolution is then best implemented as a separate hash so that we get a fast lookup for the assign-NULL case. Which feels wrong since nghttp2 already provides this functionality for us, just with this tiny glitch...

This "delayed assignment" behavior is not documented for [nghttp2_submit_request](https://nghttp2.org/documentation/nghttp2_submit_request.html). But should it really behave like that? Isn't this an nghttp2 bug?

`nghttp2_submit_request` knows which stream_id it'll become, as it returns that information just fine, and it knows the `user_data` pointer. Right?

**Comment by tatsuhiro-t** [comment]
We don't need hash because we know the address of the object.  Doubly linked list is enough in this case.

The text:

> This function returns assigned stream ID if it succeeds. But that stream is not opened yet. The application must not submit frame to that stream ID before nghttp2_before_frame_send_callback is called for this frame.

implies that stream is not created.  It says "not opened" and "not created" so I admit that I didn't write it explicitly.

> nghttp2_submit_request knows which stream_id it'll become, as it returns that information just fine, and it knows the user_data pointer. Right?

Well, yes.
I think that I can create idle stream when nghttp2_submit_request is done.  The opening of stream is still delayed, but all we want is stream object which we can assign opaque pointer.

It implies extra memory overhead if app issues lots of requests.  It is not bounded by max concurrent streams any more.  I think it would be better to add a new option to enable this new behaviour.  Thoughts?

**Comment by bagder** [comment]
> I think it would be better to add a new option to enable this new behavior. Thoughts?

(I still have not personally seen this happen so I'm basing my reasoning on @ReneReusner's discussion above with @tatsuhiro-t's input on nghttp2 details.)

> Doubly linked list is enough in this case.

Ah yes you're right. We can point directly to a node a such a linked list.

> It implies extra memory overhead

I don't think you need to allocate any more memory, I think this is just a matter of changing some logic - unless I'm misunderstanding something.

You already have the "stream => user_data" association in there in a queue somehow, set in the call to `nghttp2_submit_request()`. The problem we seem have here is that we can't clear that association with `nghttp2_session_set_stream_user_data()` immediately.

Can't we make `*set_stream_user_data()` also scan the queue of pending requests to send and clear the pending association even if the data hasn't already been sent? There's a problem when there are time gaps where we can't clear the associations. It makes them complicated to use.

This said, I'm also prepared to try out the linked-list approach. I suppose we'd then keep the list entry in the list until the `on_stream_close_callback` is called using it?

**Comment by tatsuhiro-t** [comment]
> Can't we make *set_stream_user_data() also scan the queue of pending requests to send and clear the pending association even if the data hasn't already been sent? There's a problem when there are time gaps where we can't clear the associations. It makes them complicated to use.

The problem is it might take a time because of its linear scan.
But I did it for RST_STREAM already.  The range of stream ID is bounded by the next stream ID and the stream ID at the front of the queue.  Not sure it costs real performance penalty.  It depends on how many outstanding request is queued by app.

**Comment by ReneReusner** [comment]
I don't know the design of nghttp2, so my comment below may be nonsense, but just do the simple way?

Why not removing the stream_user_data from the stream object and other queues. Then change the stream id hash map value to a 

```c
struct hash_map_entry {
    nghttp2_stream *foo;
    void *stream_user_data;
}
```

So nghttp2_submit_request will place the stream_user_data in the map but leave foo = NULL. Its assigned when the stream gets opened.

**Comment by tatsuhiro-t** [comment]
We use invasive map entry to reduce the number of allocations.  If we add another layer of objects, it incurs one more allocation per stream.

**Comment by ReneReusner** [comment]
I agree, but the discusion over on more alloc is pointless, if we need to do it anyway (a own "set" to track valid stream ids) in libcurl to fix the broken, yes its a bug, nghttp2_session_set_stream_user_data().

**Comment by bagder** [comment]
I agree with @ReneReusner. I consider this a nghttp2 "mis-feature" if not a downright bug.

We should probably consider doing the extra layer on top anyway so that we can make sure to work with existing nghttp2 installations everywhere, but I think this nghttp2 behavior should be fixed since it makes using the API really complicated.

**Comment by tatsuhiro-t** [comment]
Here is the PR to fix this in nghttp2 side as @bagder suggested
https://github.com/nghttp2/nghttp2/pull/1211

**Comment by linusnielsen** [comment]
@tatsuhiro-t It looks like your nghttp2 fix solves this issue. Thanks a lot!

Any chance that you could do an official release of nghttp2 with this fix included?

**Comment by ThomasObenaus** [comment]
Yes many thanks @tatsuhiro-t . Our test setup ran now for ~8h without a crash.

As @linusnielsen lined out it would be really helpful for us if you would be so kind to make an official release of nghttp2.
Then we can integrate the fixed version as soon as possible in our software.

**Comment by bagder** [comment]
As this problem is now fixed, even though the fix is in nghttp2, we can consider this case closed. Thank you all for participating!
